### ISSUE TITLE
Problem with dropped HTML nodes - update() doesn't work as expected

### ISSUE THREAD

**Issue by matthew-dean** [ractivejs/ractive#2272]
My experience with ractive so far has been that it performs really well with fairly flat object trees, but with more complex trees, it often tends to miss when object trees are dirty or values have changed. That is, given object A and template B, it will correctly render most of the time with updates. But then, once in a while (0.1% of the time?), given the same object structure, it will either fail to update or render certain nodes, seemingly at random. Some objects & template pairs don't seem to do this, but throughout my app, there are certain ones with these occasional failure rates, and I've formed the habit of forcing ractive to completely replace the ractive data (by creating a new object with the same structure as the old, and binding to the root data) for certain sections of the app. The ```update()``` method has no effect, no matter what keypath I give it that I know to be dirty.

Let me give you an example. I have this template:
```
{{#top}}
	{{#each .sources}}
		{{# { files: ., category: @key } }}
			<ul>
			{{#.files:i}}
				{{#if i == 0}}
					<li style="height: 23px"><h1>{{ category }}</h1></li>
				{{/if}}
				{{>file}}
			{{/}}
			</ul>
		{{/}}
	{{/each}}
{{/top}}
```
And this data:
<img width="503" alt="screen shot 2015-11-11 at 2 11 59 pm" src="https://cloud.githubusercontent.com/assets/414752/11104671/fa3d2780-887e-11e5-9b05-3c2b22757e66.png">

99.9% of the time, it renders this:
<img width="265" alt="screen shot 2015-11-11 at 2 11 12 pm" src="https://cloud.githubusercontent.com/assets/414752/11104690/0cf141ae-887f-11e5-966e-500fea9ddf5d.png">

Today, it rendered this:
<img width="260" alt="screen shot 2015-11-11 at 2 10 54 pm" src="https://cloud.githubusercontent.com/assets/414752/11104702/1aa395e0-887f-11e5-9890-606fdb722018.png">

Restarting the app, with the same data and the same template produces the correct layout with correct HTML nodes. But besides being confused and concerned about why this happens and how often it might happen, update() has no effect, even though the documentation states that it will do a dirty check. Well... I should say, that is, it has no effect on rendering the correct HTML nodes. If I use ```ractive.get()``` it shows the above object tree. Both *"CSS"* and *"LESS"* are there, present, in the ractive object tree, but, in this 0.1% case, it drops an HTML node (and obviously all of its child nodes).

I can tell you with certainty as I've developed with Ractive over the last year, that while objects and templates don't change their structure, the resulting HTML is not 100% predictable. Like I said, the failure rate is fairly low (and there are never errors in the console), but in some cases, given Object A and Template B, you will get Layout X 99.9% of the time, and Layout Y 0.1% of the time. I don't know if it's an issue in the HTML diffing algorithm, but  part of my question (other than the obvious "why is this happening") - is there a way to tell Ractive to always ALWAYS make sure that the data it has corresponds to the HTML it's rendering? As in, even if it THINKS the data hasn't changed, for whatever reason, that it really super serious has created the nodes that it thinks it has? And if that can't be done via ```update()```, what would be the correct method?

One thing I should say about this: once a Ractive has failed, it will always fail. That is, if you keep rebinding an object, even if you use ```set()``` and ```update()``` at various keypaths, it will forever omit the HTML nodes it is missing (unless, I think, the object structure has changed significantly? I haven't figured out a pattern here).

I almost think that it's Ractive being overly conservative here - trying to only render the HTML it needs to based on data changes. But I wonder if Ractive shouldn't check its work afterwards, if it's not going to update all the HTML. But I don't know the lib well enough to suggest how to do this. (Would this be ```reset()```? Or the undocumented ```resetTemplate()```?)

As this pertains to core functionality / purpose of Ractive, I'm wondering if you can help me out. Is there something I may have missed? What might cause these occasional failures? And how can they be avoided?

**Comment by heavyk** [comment]
have you tried this with edge?

**Comment by matthew-dean** [comment]
@heavyk No, this is with the latest official release. Why, is there reason to believe this wouldn't happen with 0.8? The problem for me is that I'm close to releasing the software where this is used, so I would be hesitant to change versions right before a final release. And it happens so infrequently that it's hard to tell if it's solved or not. Which I know makes this a ¯\\_(ツ)_/¯ kind of bug, I know. But I swear to you that this is happening as described.

**Comment by matthew-dean** [comment]
Maybe the question could be phrased this way, if the source of the problem is not known -- Is there a way to tell Ractive to always entirely create the virtual DOM when values change in the data, and check that against rendered nodes to see if anything is missing? Without ruining Ractive performance?

**Comment by matthew-dean** [comment]
This happened again today. This time, neither CSS nor LESS nodes were rendered. Same object structure. Using ```ractive.get()``` demonstrated that the data was present, just not rendered.

I tried the equivalent of ```ractive.set('top', ractive.get('top'))``` and ```ractive.set('top', JSON.parse(JSON.stringify(ractive.get('top'))))```.

I inspected the ractive.template object structure to see if maybe the parsing was mucked (which wouldn't really make sense since it would be unchanged during the same session. Still nothing. Template is there and parsed, and data is there, it just doesn't render some of the time. Somehow ractive spontaneously stops knowing how to render the data.

**Comment by martypdx** [comment]
@matthew-dean Sorry you're having issues.

Having just built a fairly large app using `0.7` this year, I haven't noticed sporadic failures as you mention. But the two areas I had problems with were transitions in nested and dynamic components, and I noticed something similar to what you mentioned with nested `if` blocks within components that involved mapped (from higher level parent) data.

I could even put in a `{{console.log(someData)}}` and see that expression being re-evaluated, but an adjacent `if` block would just not update.

I worked around this by unnesting the blocks. So instead of:

```
{{#if foo}}
    foo
    {{#if bar}}bar{{/bar}}{{/if}}
    qux
{{/if}}
```

I changed it to:

```
{{#if foo}}foo{{/if}}
{{#if foo && bar}}bar{{/bar}}{{/if}}
{{#if foo}}qux{{/if}}
```

I have to say that I have found Ractive to be very deterministic. I suspect because it updates synchronously. So when I saw this condition, I could always recreate it reliably. 

Just looking for what would be causing the difference in working vs not working: Are you doing any direct DOM manipulation? Or is there asynchronicity that may be resolving at different times? Are there particular data values for which this is happening?

Given what you have described, I suspect under the covers that either the model layer or the virtual believes the data has already changed and so calling `update` won't help (as you have noticed). Ideally we would resolve the bugs, but perhaps it would be useful to have a `ractive.update( 'foo', { force: true } )` option that would force the dependants to update completely.

**Comment by matthew-dean** [comment]
@martypdx I've found the issue! And can reproduce it!

Basically, once you set a keypath to undefined, Ractive 0.7.3 will never update that keypath, ever again. At least with an each, but possibly with other bindings. (Which is obviously not right, since it should be the same behavior as Handlebars, where an "each" can be shorthand for "forEach only if not falsy".) And, funnily enough, I had just tried the "if" workaround which worked for me.

0.7.3 bug demonstration: http://codepen.io/matthewdean/pen/epbvjy
0.7.3 bug workaround with if: http://codepen.io/matthewdean/pen/yYGbNK
0.8-edge (fixed) http://codepen.io/matthewdean/pen/ZbVKbR

My last link verifies that this is fixed in latest edge version. However, I couldn't find any previous issue that describes this bug, so maybe this should be marked as a fixed bug? And hopefully there's now a test for this in 0.8?  And also for your test cases, because working "if" and "each" blocks is kind of a fundamental building block of ractive templates.

This has been driving me crazy so I'm at least glad there are ways around it.
